devicetypes/smartthings/zwave-multi-metering-switch.src/zwave-multi-metering-switch.groovy

```diff
@@ -0,0 +1,267 @@
+		}
+
+		main(["switch"])
+		details(["switch","power","energy","refresh","reset"])
+	}
+}
+
+def parse(String description) {
+	def result = null
+	if (description.startsWith("Err")) {
+		result = createEvent(descriptionText:description, isStateChange:true)
+	} else if (description != "updated") {
+		def cmd = zwave.parse(description)
+		if (cmd) {
+			result = zwaveEvent(cmd, null)
+		}
+	}
+	log.debug "parsed '${description}' to ${result.inspect()}"
+	result
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.multichannelv3.MultiChannelCmdEncap cmd, ep) {
+	log.debug "Multichannel command ${cmd}" + (ep ? " from endpoint $ep" : "")
+	def encapsulatedCommand = cmd.encapsulatedCommand()
+	zwaveEvent(encapsulatedCommand, cmd.sourceEndPoint as Integer)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.basicv1.BasicReport cmd, ep) {
+	log.debug "Basic ${cmd}" + (ep ? " from endpoint $ep" : "")
+	def value = cmd.value ? "on" : "off"
+	ep ? changeSwitch(ep, value) : []
+}
+
+private changeSwitch(endpoint, value) {
+	def result = []
+	if(endpoint == 1) {
+		result += createEvent(name: "switch", value: value, isStateChange: true, descriptionText: "Switch ${endpoint} is ${value}")
+	} else {
+		String childDni = "${device.deviceNetworkId}/$endpoint"
+		def child = childDevices.find { it.deviceNetworkId == childDni }
+		child?.sendEvent(name: "switch", value: value, isStateChange: true, descriptionText: "Switch ${endpoint} is ${value}")
+	}
+	result
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.meterv3.MeterReport cmd, ep) {
+	log.debug "Meter ${cmd}" + (ep ? " from endpoint $ep" : "")
+	def result
+	if(ep == 1) {
+		result = handleMeterReport(cmd)
+	} else if(ep) {
+		result = childHandleMeterReport(cmd, ep)
+	} else {
+		result = zwaveEvent(cmd)
+	}
+	result
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.meterv3.MeterReport cmd) {
+	def event = createEvent([isStateChange:  false, descriptionText: "Wattage change has been detected. Refreshing each endpoint"])
+	[event, response(refreshAll())]
+}
+
+def handleMeterReport(cmd) {
+	def result = []
+	if (cmd.meterType == 1) {
+		if (cmd.scale == 0) {
+			result += createEvent(name: "energy", value: cmd.scaledMeterValue, unit: "kWh")
+			result += response(encap(zwave.meterV3.meterGet(scale: 2), 1))
+		} else if (cmd.scale == 2) {
+			result += createEvent(name: "power", value: Math.round(cmd.scaledMeterValue), unit: "W")
+		}
+	}
+	result
+}
+
+def childHandleMeterReport(cmd, endpoint) {
+	String childDni = "${device.deviceNetworkId}/$endpoint"
+	def child = childDevices.find { it.deviceNetworkId == childDni }
+	def result = []
+	if (cmd.meterType == 1) {
+		if (cmd.scale == 0) {
+			child?.sendEvent(name: "energy", value: cmd.scaledMeterValue, unit: "kWh")
+			result += response(encap(zwave.meterV3.meterGet(scale: 2), endpoint))
+		} else if (cmd.scale == 2) {
+			child?.sendEvent(name: "power", value: Math.round(cmd.scaledMeterValue), unit: "W")
+		}
+	}
+	result
+}
```

Comment:
you could avoid the code duplication here by creating one method that just creates all of the events you need after a meter report. The initial handler above could be responsible for selecting whether the parent or child dispatches them.